src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -39,118 +28,550 @@ const LINE_SHAPE_WIDTH   : f32 = LINE_WIDTH + 2.0 * PADDING;
+            }
+
+            fn normal_local(&self, point:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+                point_rotation(point)
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                let radius = self.shape.radius.get();
+                let center = Vector2::zero();
+                let point_to_center = point.xy() - center;
+
+                let closest_point = center + point_to_center / point_to_center.magnitude() * radius;
+                let vector_angle  = -nalgebra::Rotation2::rotation_between(&Vector2::new(0.0, 1.0),&closest_point).angle();
+                let start_angle   =  self.shape.start_angle.get();
+                let end_angle     =  start_angle + self.shape.angle.get();
+                let upper_bound   = start_angle.max(end_angle);
+                let lower_bound   = start_angle.min(end_angle);
+
+                let correct_quadrant = lower_bound < vector_angle && upper_bound > vector_angle;
+                if correct_quadrant {
+                     Some(Vector2::new(closest_point.x, closest_point.y))
+                } else {
+                    None
+                }
             }
         }
     }
 }}
 
-macro_rules! define_line {($($color:tt)*) => {
+macro_rules! define_line {($color:expr, $highlight_color:expr) => {
     /// Shape definition.
     pub mod line {
         use super::*;
         ensogl::define_shape_system! {
-            () {
+            (focus_split_center:Vector2<f32>, focus_split_angle:f32) {
                 let width  = LINE_WIDTH.px();
                 let height : Var<Pixels> = "input_size.y".into();
-                let shape  = Rect((width,height));
-                let shape  = shape.fill(color::Rgba::from($($color)*));
-                shape.into()
+                let shape  = Rect((width.clone(),height));
+
+                let split_shape = FocusedEdge::new(
+                    shape,&focus_split_center.px(),&focus_split_angle.into());
+                let shape       = split_shape.fill($color, $highlight_color);
+                hover_area(shape,HOVER_EXTENSION.px()).into()
+            }
+        }
+
+        impl EdgeShape for component::ShapeView<Shape> {
+            fn set_focus_split_center_local(&self, center:Vector2<f32>) {
+                self.shape.focus_split_center.set(center);
+            }
+
+            fn set_focus_split_angle(&self, angle:f32) {
+                self.shape.focus_split_angle.set(angle);
+            }
+
+            fn sprite(&self) -> &Sprite{
+                &self.shape.sprite
+            }
+
+            fn events(&self) -> &ShapeViewEvents{
+                &self.events
+            }
+
+            fn normal_local(&self, _:Vector2<f32>) -> nalgebra::Rotation2<f32> {
+                nalgebra::Rotation2::new(0.0)
+            }
+
+            fn snap_local(&self, point:Vector2<f32>) -> Option<Vector2<f32>> {
+                #[allow(deprecated)]
```

Comment:
all such places needs to be propelry documented. No one will remember to fix that in 1 month from now without proper documentation and link to an issue.